flat_mutation_reader.cc

```diff
@@ -912,3 +915,47 @@ class generating_reader final : public flat_mutation_reader::impl {
 flat_mutation_reader make_generating_reader(schema_ptr s, std::function<future<mutation_fragment_opt> ()> get_next_fragment) {
     return make_flat_mutation_reader<generating_reader>(std::move(s), std::move(get_next_fragment));
 }
+
+bool mutation_fragment_stream_validator::operator()(const dht::decorated_key& dk) {
+    // FIXME: Validate fragment position monotonicity
+    //        https://github.com/scylladb/scylla/issues/4804
+    return true;
+}
+
+template<typename T>
+sstring position_desc(mutation_fragment::kind kind, const T& pos) {
+    return format("{}{}", kind == mutation_fragment::kind::range_tombstone ? "range_tombstone: " : "", pos);
+};
+
+mutation_fragment_stream_validator::~mutation_fragment_stream_validator() {
+    if (!_prev_pos.is_partition_end()) {
+        on_internal_error(fmr_logger, format("Stream ended with unclosed partition: {}", position_desc(_prev_kind, _prev_pos)));
+    }
+}
+
+bool mutation_fragment_stream_validator::operator()(const mutation_fragment& mv) {
+    auto kind = mv.mutation_fragment_kind();
+    auto pos = mv.position();
+    bool valid;
+
+    fmr_logger.debug("{} -> {}", position_desc(_prev_kind, _prev_pos), position_desc(kind, pos));
```

Comment:
`position_desc()` will be invoked even if the log level is disabled, which adds performance overhead. We should only pass printable objects here.

Reply:
An alternative is to use `seastar::value_of()` which is evaluted only if the log line is actually printed.

Reply:
Or maybe surround with `if (fmr_logger.is_enabled(log_level::debug))` but that might be too low level.

Reply:
And as @tgrabiec suggested if we pass printable objects then their operator<<  will be called only if the log line is being printed.